packages/core/factories/src/useUserAddressFactory.ts

```diff
@@ -0,0 +1,97 @@
+import { UseUserAddress, UserAddressType, UseUser } from '@vue-storefront/interfaces';
+import { Ref, ref, computed } from '@vue/composition-api';
+
+export interface UpdatedUserAddresses<ADDRESS, UPDATE_USER_PARAMS> {
+  addresses: ADDRESS[];
+  updateUserParams?: UPDATE_USER_PARAMS;
+}
+
+export type UseUserAddressFactoryParams<
+  USER,
+  ADDRESS,
+  ADDRESS_TYPE = UserAddressType,
+  SEARCH_PARAMS = { [x: string]: any },
+  UPDATE_USER_PARAMS = any,
+> = {
+  addresses: ADDRESS[];
+  searchAddresses: (params?: SEARCH_PARAMS) => Promise<ADDRESS[]>;
+  addAddress: (address: ADDRESS, type: ADDRESS_TYPE) => Promise<UpdatedUserAddresses<ADDRESS, UPDATE_USER_PARAMS>>;
+  deleteAddress: (address: ADDRESS) => Promise<UpdatedUserAddresses<ADDRESS, UPDATE_USER_PARAMS>>;
+  updateAddress: (address: ADDRESS) => Promise<UpdatedUserAddresses<ADDRESS, UPDATE_USER_PARAMS>>;
+  getBillingAddresses: () => ADDRESS[];
+  getShippingAddresses: () => ADDRESS[];
+  userComposable?: UseUser<USER, UPDATE_USER_PARAMS>;
+}
+
+export function useUserAddressFactory<USER, ADDRESS, ADDRESS_TYPE = UserAddressType, SEARCH_PARAMS = { [x: string]: any }, UPDATE_USER_PARAMS = any> (factoryParams: UseUserAddressFactoryParams<USER, ADDRESS, ADDRESS_TYPE, SEARCH_PARAMS>) {
+  const loading: Ref<boolean> = ref(false);
+
+  const updateData = async (data: UpdatedUserAddresses<ADDRESS, UPDATE_USER_PARAMS>) => {
+    factoryParams.addresses.length = 0;
+    data.addresses.forEach(address => factoryParams.addresses.push(address));
+
+    if (factoryParams.userComposable && data.updateUserParams) {
+      await factoryParams.userComposable.updateUser(data.updateUserParams);
+    }
+
+    loading.value = false;
+  };
+
+  const logError = (error: any) => {
+    console.log(error);
+    loading.value = false;
+
+    throw error;
+  };
+
+  const searchAddresses = async (params) => {
+    loading.value = true;
+
+    const addresses = await factoryParams.searchAddresses(params);
+
+    updateData({ addresses });
+
+    loading.value = false;
+  };
+
+  const getBillingAddresses = () => factoryParams.getBillingAddresses();
+
+  const getShippingAddresses = () => factoryParams.getShippingAddresses();
+
+  const addAddress = async (address: ADDRESS, type: ADDRESS_TYPE) => {
+    loading.value = true;
+
+    await factoryParams.addAddress(address, type)
+      .then(updateData)
+      .catch(logError);
```

Comment:
Addressing the comment above, managing the `loading` state should be more transparent. 
Especially in the context of upcoming changes that will introduce dividing this `loading` to several flags grouped under this on name. 
```suggestion
    try {
      const result = await factoryParams.addAddress(address, type);
      updateData(result);
    } catch (err) {
      console.error(err);
      throw err;
    } finally {
      loading.value = false;
    }
```

Reply:
I don't like this approach. I prefer DRY. Especially it works the same